### PR TITLE
Vararg completion

### PR DESCRIPTION
**Pull request by rcorre**
Offer successive completion for varargs.

Right now, this applies to `config-cycle`. Instead of stopping after the first value is entered, it will repeatedly suggest more values.

This should also support completing successive tags for the future `bookmark-tag` and `bookmark-load` commands for #882.

I also think there was some other feature that was blocked by the completion not regenerating as args change, but I can't find the conversation any more.

---
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/3790)

### CODE REVIEW

File: qutebrowser/commands/command.py [hunk 1]
```diff
@@ -170,6 +171,8 @@ def get_arg_info(self, param):
 
     def get_pos_arg_info(self, pos):
         """Get an ArgInfo tuple for the given positional parameter."""
+        if pos >= len(self.pos_args) and self._has_vararg:
+            pos = len(self.pos_args) - 1
```
Comment (jgkamat) [inline qutebrowser/commands/command.py]: A bit of a nitpick, but this can be:

```
if self._has_vararg:
  pos = min(len(self.pos_args) - 1, pos))
```

Comment (rcorre) [reply qutebrowser/commands/command.py]: I think the current way is a little more readable

**Comment by rcorre** [comment]
@The-Compiler whenever you get a chance, could you let me know if this looks good to merge? The [one failure](https://travis-ci.org/qutebrowser/qutebrowser/jobs/360738448#L922) is in `test_stylesheet`, it seems unrelated.

**Comment by rcorre** [comment]
> Why does this cause the completion to be regenerated?

Previously the completer only instantiated a new completion model if the completion function changed. Now it instantiates a new completion model whenever the arg list changes. For example:

| command text                  | old call | new call|
| ------------------------------| -------- | ------- |
| `config-cycle \|`             | `configmodel.option()`              | `configmodel.option()`             | 
| `config-cycle opt \|`         | `configmodel.value()`                | `configmodel.value()`               | 
| `config-cycle opt foo \|`     | | `configmodel.value('foo')`          | `configmodel.value('foo')`         |
| `config-cycle opt foo bar\|`  | | | 
| `config-cycle opt foo bar \|` | | `configmodel.value('foo', 'bar')` | 

> When is it regenerated exactly?

Whenever the arg list changes. This does _not_ include changes to the word the cursor is on, that just changes the pattern. As soon as you press space, the completion will be regenerated.

> What are the performance implications of doing so?

Potentially a longer wait every time you complete typing a word. However, most completion functions are really fast. The only risky one is the `url` model, and due to `maxsplit` everything after `open` is part of the same arg, so it won't regenerate any more often than it used to.

> Why are those "deduplication" changes in the value configmodel needed?

Previously `config-cycle` would only complete a value once. Now it offers completion for successive values. However, I didn't think it made much sense to enter the same value twice, so once the user has entered a value it will be removed from the completion. For example, if you've typed `config-cycle backend webengine |`, it will only offer `webkit` as a completion for the next word.

For the upcoming `bookmark-tag` I use the same logic to filter out tags the user has already input, since it doesn't make sense to apply the same tag twice. If we don't care about deduping we can actually return to the same regeneration logic, but we may want to regenerate whenever the args change if we ever want to support flag-aware completion.

File: qutebrowser/commands/command.py (hunk 1 above)
Comment (mschilli87) [reply qutebrowser/commands/command.py]: Just to join the nitpicking, my preferred version would be

```python
if self._has_vararg  and pos >= len(self.pos_args):
    pos = len(self.pos_args) - 1
```
:wink:

**Comment by The-Compiler** [comment]
And finally merged - thanks!

**Pull request merged**